packages/aws-cdk/lib/version.ts

```diff
@@ -0,0 +1,98 @@
+import { exec as _exec } from 'child_process';
+import colors = require('colors/safe');
+import { open as _open, stat as _stat } from 'fs';
+import semver = require('semver');
+import { promisify } from 'util';
+import { debug, print, warning } from '../lib/logging';
+
+const ONE_DAY_IN_SECONDS = 1 * 24 * 60 * 60;
+
+const exec = promisify(_exec);
+const open = promisify(_open);
+const stat = promisify(_stat);
+
+export const DISPLAY_VERSION = `${versionNumber()} (build ${commit()})`;
+
+function versionNumber(): string {
+  return require('../package.json').version.replace(/\+[0-9a-f]+$/, '');
+}
+
+function commit(): string {
+  return require('../build-info.json').commit;
+}
+
+export class CacheFile {
+  private readonly file: string;
+
+  // File modify times are accurate only till the second, hence using seconds as precision
+  private readonly ttlSecs: number;
+
+  constructor(file: string, ttlSecs: number) {
+    this.file = file;
+    this.ttlSecs = ttlSecs;
+  }
+
+  public async hasExpired(): Promise<boolean> {
+    try {
+      const lastCheckTime = (await stat(this.file)).mtimeMs;
+      const today = new Date().getTime();
+
+      if ((today - lastCheckTime) / 1000 > this.ttlSecs) { // convert ms to secs
+        return true;
+      }
+      return false;
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        return true;
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  public async update(): Promise<void> {
+    await open(this.file, 'w');
+  }
+}
+
+// Export for unit testing only.
+// Don't use directly, use displayVersionMessage() instead.
+export async function latestVersionIfHigher(currentVersion: string, cacheFile: CacheFile): Promise<string | null> {
+  if (!(await cacheFile.hasExpired())) {
+    return null;
+  }
+
+  const { stdout, stderr } = await exec(`npm view cdk version`);
+  if (stderr && stderr.trim().length > 0) {
+    debug(`The 'npm view' command generated an error stream with content [${stderr.trim()}]`);
```

Comment:
Might as well return here, what are the odds `stdout` is still going to contain a version number?

Reply:
Classic case -- Not connected to the internet. `stdout` has the latest version cached from the last repo sync and `stderr` has the corresponding warning.